architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+
+```rust
+/// Configuration for computing and distributing rewards to validators
+pub struct RewardConfig {
+    /// Rewards pool cap (`monetary_expansion_cap`)
+    cap: Coin,
+    /// Upper bound for reward distribution rate (`monetary_expansion_r0`)
+    r0: FixedNumber,
+    /// Tau (`monetary_expansion_tau`)
+    tau: FixedNumber,
+    /// Decay factor for tau (`monetary_expansion_decay`)
+    decay_factor: u32,
+    /// Period after which rewards are distributed (`distribution_period` in secs)
+    distribution_period: u32,
+}
+
+impl RewardConfig {
+    /// Calculates reward for given values
+    pub fn calculate_reward(&self, sum_average_stake: Coin, minted: Coin) -> Coin {
+        todo!()
+    }
+
+    /// Decays `tau` based on `decay_factor`
+    pub fn decay_tau(&mut self) {
+        todo!()
+    }
+}
+```
+
+`RewardsPoolState` struct tracks already distributed rewards and calculates rewards during distribution.
+
+```rust
+/// Tracks and computes rewards for validators
+pub struct RewardPoolState {
+    /// Reward configuration (from genesis)
+    pub config: RewardConfig,
+    /// Rewards accumulated in current period
+    pub period_bonus: Coin,
+    /// last block height that updated it (i64 from Tendermint protobuf)
+    pub last_block_height: BlockHeight,
+    /// last reward distribution time
+    pub last_distribution_time: Timespec,
+    /// Record how many coins have been minted, can't exceed the cap
+    pub minted: Coin,
+}
+
+impl RewardPoolState {
+    /// Calculates reward for given `sum_average_stake`
+    ///
+    /// # Note
+    ///
+    /// `sum_average_stake` can be calculated using `ProposerTracker::get_sum_average_stake()`
+    #[inline]
+    pub fn calculate_reward(&self, sum_average_stake: Coin) -> Coin {
+        self.config.calculate_reward(total_stake, self.minted)
+    }
+
+    /// Decays `tau` based on `decay_factor` in rewards config
+    #[inline]
+    pub fn decay_tau(&mut self) {
+        self.config.decay_tau()
+    }
+
+    /// Updates rewards pool state with given values
+    pub fn update(
+        &mut self, 
+        period_bonus: Coin, 
+        last_block_height: BlockHeight, 
+        last_distribution_time: Timespec, 
+        minted: Coin,
+    ) {
+        self.period_bonus = period_bonus;
+        self.last_block_height = last_block_height;
+        self.last_distribution_time = last_distribution_time;
+        self.minted = minted;
+    }
+}
+```
+
+`ProposerTracker` tracks the number of blocks proposed by any validator and also maintains all the validator set changes
+in current reward period.
+
+```rust
+/// Tracks the number of blocks proposed by validators and also their voting power changes
+pub struct ProposerTracker {
+    /// Number of blocks a validator proposed in current reward period
+    proposer_count: BTreeMap<StakedStateAddress, u64>,
+    /// Changes in active validator set in current reward period
+    validator_set_changes: Vec<(Timespec, Vec<Validator>)>,
```

Comment:
what is this one for @lezzokafka @yihuang @devashishdxt ? In https://crypto-com.github.io/getting-started/staking.html#rewards -- it says "number of blocks proposed by the validator / total number of blocks"

Reply:
This is to solve second point mentioned in `Problems with current design`. We should consider all the changes in active validator set when calculating total stake.

Reply:
> For example, when a validator unbonds or deposits funds just before rewards distribution (let us assume that they unbonded their funds at 23rd hour), their stake is not considered in rewards calculation.

I guess that's a feature / can be specified this way (if it makes things less complex and faster)

Reply:
If we want that behaviour, then I guess this isn’t required. But, I think we should add this averaging to calculate total stake, this’ll be more _fair_ for validators. I discussed this with @lezzokafka before leaving, he may be able to explain motivation behind this.

Reply:
@devashishdxt  @tomtau  Sure, the current way of calculating the total reward ("minted") being sent to the pool is based on the _daily_reward_rate * total_staking_ at the cut off time.
For example: 
- a huge validator joins the validator set at the very last second; the reward being sent to the pool will be calculated by the _total_staking_ at the cut off time.